packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,290 @@
+pragma solidity ^0.8.0;
+
+import "./MultiInvoker.sol";
+
+// import "forge-std/Test.sol";
+
+contract MultiInvokerRollup is MultiInvoker {
+
+    event AddressAddedToCache(address indexed addr, uint256 nonce);
+
+    uint256 public addressNonce;
+    mapping(uint => address) public addressCache;
+    mapping(address => uint) public addressNonces;
+
+    
+    constructor(
+        Token6 usdc_,
+        IBatcher batcher_,
+        IEmptySetReserve reserve_,
+        IController controller_
+    ) MultiInvoker(
+        usdc_,
+        batcher_,
+        reserve_,
+        controller_
+    ) { }
+
+    /// @dev fallback eliminates the need to include function sig
+    fallback (bytes calldata input) external returns (bytes memory) {
+        decodeFallbackAndInvoke(input);
+    }
+
+    /// @dev this function serves (@todo will serve*) exactly the same as invoke(Invocation[] memory invocations),
+    /// but includes all the logic to handle the highly packed calldata
+    function decodeFallbackAndInvoke(bytes calldata input) private {
+        uint ptr;
+        uint len = input.length;
+    
+        for(ptr; ptr < len;) {
+            uint8 action = toUint8(input[ptr:ptr+1]);
+            ptr += 1;
+
+            // solidity doesn't like evaluating bytes as enums :/ 
+            if(action == 1) { // DEPOSIT
+                address account; address product; UFixed18 amount;
+                (account, product, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                depositTo(account, IProduct(product), amount);
+            } else if (action == 2) { // WITHDRAW
+                address receiver; address product; UFixed18 amount;
+                (receiver, product, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                collateral.withdrawFrom(msg.sender, receiver, IProduct(product), amount);
```

Comment:
can we pull these non-helper-function pieces of logic into internal helper functions as well?

now that we're calling these from two places, there's danger that we may have two different implementation of a clause. we should consolidate these into callable internal functions in the multi-invoker.

Reply:
like for collateral/product? yes would add in MultiInvoker.sol with `_depositTo()` etc

Reply:
theres a conflict because `wrap()` and `unwrap()` both call `_wrap()` and `_unwrap()` respectively. this is annoying because it hurts the continuity of this style. possible to consolidate the functions instead of breaking the `_` rule?

Reply:
otherwise everything is compiling and all unit tests passing

Reply:
how about we do:

`wrap / unwrap` -> `_wrap / _unwrap`
and
`_wrap / _unwrap` -> `_handleWrap / _handleUnwrap`

The `_` is very important because it lets us quickly identify which functions can be called via external transaction, so we don't want to compromise on that one.

Reply:
cool adding